### PR TITLE
[vNext Tokens] Re-tokenizing TabBar

### PR DESCRIPTION
**Pull request by edjamesmsft**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

This change demonstrates the new strategy for handling tokenization of compound controls, where the "contained" control(s) is/are internal only. We expose only a small subset of tokens of the internal control to the public, via special nil-able values in the containing control's token set, which then sets the internal, private control's token set via a custom derived token class. This allows the public to change a subset of tokens, while still allowing our token system, and hence designers, to still have full control and access to all tokens of the internal class.

### Verification

Manually tested via Simulator.

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| TokenizeTabBarAfter | ReTokenizeTabBar |

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/990)

### CODE REVIEW

File: ios/FluentUI.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 1]
```diff
@@ -1,23 +1,25 @@
 {
-  "pins" : [
-    {
-      "identity" : "appcenter-sdk-apple",
-      "kind" : "remoteSourceControl",
-      "location" : "https://github.com/microsoft/appcenter-sdk-apple.git",
-      "state" : {
-        "revision" : "25f64229373de97ff3920941cd52203193e5d8be",
-        "version" : "4.3.0"
+  "object": {
```
Comment (huwilkes) [inline ios/FluentUI.xcworkspace/xcshareddata/swiftpm/Package.resolved:2]: I've also had this file changing a bunch of me recently. Not sure if we need to update it, but if we do it should probably be in a different PR.

File: ios/FluentUI/Tab Bar/TabBarTokens.swift [hunk 2]
```diff
@@ -17,66 +17,14 @@ open class TabBarTokens: ControlTokens {
...
-    open var badgeCornerRadii: CGFloat { 10.0 }
+    open var tabBarItemUnselectedColor: DynamicColor? { nil }
 
     /// Font info for the title label when in portrait view
-    open var titleLabelPortrait: FontInfo { return .init(size: 10, weight: .medium) }
+    open var tabBarItemTitleLabelPortrait: FontInfo? { nil }
 
     /// Font info for the title label when in landscape view
-    open var titleLabelLandscape: FontInfo { return .init(size: 13, weight: .medium) }
+    open var tabBarItemTitleLabelLandscape: FontInfo? { nil }
```
Comment (huwilkes) [inline ios/FluentUI/Tab Bar/TabBarTokens.swift:29]: Can we add the word "font" to `tabBarItemTitleLabelPortrait` and `tabBarItemTitleLabelLandscape`? Those names sound like the actual strings/labels rather than their fonts to me.

File: ios/FluentUI/Tab Bar/TabBarItemTokens.swift [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    open var badgeBorderWidth: CGFloat { globalTokens.borderSize[.thick] }
+
+    /// The horizontal padding for the `BadgeLabel`
+    open var badgeHorizontalPadding: CGFloat { 10.0 }
+
+    /// The radii of the four corners of the `BadgeLabel`
+    open var badgeCornerRadii: CGFloat { 10.0 }
+
+    /// Font info for the title label when in portrait view
+    open var titleLabelPortrait: FontInfo { return .init(size: 10, weight: .medium) }
```
Comment (mischreiber) [inline ios/FluentUI/Tab Bar/TabBarItemTokens.swift:69]: Same question about scaling or non-scaling for these values.

File: ios/FluentUI/Tab Bar/TabBarView.swift [hunk 4]
```diff
@@ -191,13 +191,43 @@ open class TabBarView: UIView, TokenizedControlInternal {
         }
     }
 
+    private class CustomTabBarItemTokens: TabBarItemTokens {
+        var tabBarTokens: TabBarTokens = .init()
```
Comment (mischreiber) [inline ios/FluentUI/Tab Bar/TabBarView.swift:195]: I don't think we should be initializing this to a default value here. Presumably that will cause the `required init()` below to fail, but we can work around that by having it throw an error (since no one should be calling it anyway):

```swift
@available(*, unavailable)
required init() {
    preconditionFailure("init() has not been implemented")
}
```

File: ios/FluentUI.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 1 above)
Comment (edjamesmsft) [reply ios/FluentUI.xcworkspace/xcshareddata/swiftpm/Package.resolved:2]: I can remove it from this pr at the very least

File: ios/FluentUI/Tab Bar/TabBarTokens.swift (hunk 2 above)
Comment (edjamesmsft) [reply ios/FluentUI/Tab Bar/TabBarTokens.swift:29]: Will do!

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift [hunk 5]
```diff
@@ -237,27 +237,11 @@ extension TabBarViewDemoController: DemoAppearanceDelegate {
...
-        }
-
-        override var phonePortraitHeight: CGFloat {
-            return 35.0
-        }
-
-        override var phoneLandscapeHeight: CGFloat {
-            return 35.0
+        override var tabBarItemTitleLabelPortrait: FontInfo? {
+            return .init(size: 15, weight: .bold)
```
[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift:241]: Who would I verify with? Design? PM?

[the remark this answers is not in the document]
Comment (harrieshin) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift:241]: what does apple native app do for text? i hunch would be it will increase upto somepoint but stop and use larger context viewer for really big fonts.

**Review by mischreiber — APPROVED** [review]

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift:241]: Followed up offline: these should probably be non-scaling

[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift:241]: These are already shouldScale: false where the FontInfo is actually consumed in TabBarItemView.swift, line 261 and 269. Is that sufficient?

[the remark this answers is not in the document]
Comment (edjamesmsft) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/TabBarViewDemoController.swift:241]: Followed up offline, yes it is sufficient!
